src/Microsoft.Azure.EventHubs/EventHubsDiagnosticSource.cs

```diff
@@ -0,0 +1,125 @@
+
+            // extract activity tags from input
+            activity.AddTag("component", "Microsoft.Azure.EventHubs");
+            activity.AddTag("span.kind", "producer");
+            activity.AddTag("operation.name", $"Send");
+            activity.AddTag("operation.data", $"{csb.EntityPath}/{partitionKey}");
+            activity.AddTag("peer.service", "Azure Event Hub");
+            activity.AddTag("peer.hostname", csb.Endpoint.OriginalString);
+            activity.AddTag("eh.event_hub_name", csb.EntityPath);
+            activity.AddTag("eh.partition_key", partitionKey);
+            activity.AddTag("eh.event_count", count.ToString());
+            activity.AddTag("eh.client_id", clientId);
+
+            // in many cases activity start event is not interesting, 
+            // in that case start activity without firing event
+            if (DiagnosticListener.IsEnabled(SendActivityStartName))
+            {
+                DiagnosticListener.StartActivity(activity,
+                    new
+                    {
+                        Endpoint = csb.Endpoint,
+                        EntityPath = csb.EntityPath,
+                        PartitionKey = partitionKey,
+                        EventDatas = eventDatas
+                    });
+            }
+            else
+            {
+                activity.Start();
+            }
+
+            return activity;
+        }
+
+        internal static void FailSendActivity(Activity activity, EventHubsConnectionStringBuilder csb, string partitionKey, IEnumerable<EventData> eventDatas, Exception ex)
+        {
+            // TODO consider enriching activity with data from exception
+
+            if (!IsEnabled || !DiagnosticListener.IsEnabled(SendActivityExceptionName))
+            {
+                return;
+            }
+
+            DiagnosticListener.Write(SendActivityExceptionName,
+                new
+                {
+                    Endpoint = csb.Endpoint,
+                    EntityPath = csb.EntityPath,
+                    PartitionKey = partitionKey,
+                    EventDatas = eventDatas,
+                    Exception = ex
+                });
+        }
+
+        internal static void StopSendActivity(Activity activity, EventHubsConnectionStringBuilder csb, string partitionKey, IEnumerable<EventData> eventDatas, Task sendTask)
+        {
+            if (activity == null)
+            {
+                return;
+            }
+
+            // stop activity
+            activity.AddTag("error", (sendTask?.Status == TaskStatus.RanToCompletion).ToString());
+            DiagnosticListener.StopActivity(activity,
+                new
+                {
+                    Endpoint = csb.Endpoint,
+                    EntityPath = csb.EntityPath,
+                    PartitionKey = partitionKey,
+                    EventDatas = eventDatas,
+                    TaskStatus = sendTask?.Status
```

Comment:
How about "Succeeded" instead of "TaskStatus"? Might not be obvious what "Task" refers to?!

Reply:
So this is still under discussion for the guidance that we are working on but we would like to standardize some payload fields, like "Exception" and "TaskStatus". 

The idea for payload is to pass raw objects (not formatted/parsed) so that custom listeners can extract whatever they want from them. Tags contain more simplified information prepared for efficient consumption (e.g. tracing)

Reply:
@cwe1ss 
I agree that TaskStatus name is not the best one.

However Succeeded implies binary result. TaskStatus (with proper name) in reality has 3: success, failure, cancellation